dv/uvm/icache/doc/ibex_icache_dv_plan.md

```diff
@@ -1,99 +1,142 @@
+
+![Agent connections](agent-connections.svg)
+
+All three agents report events to the scoreboard.
+The core agent reports every branch sent to the cache, invalidation requests, enabling/disabling the cache and every instruction fetched (address and contents).
+The memory agent reports every change of seed (see the [Memory Agent](#memory-agent) section below).
+The busy agent reports every edge of the busy signal.
+
+### Agents
+
+#### Core Agent
+
+The core agent emulates a core making requests of the cache.
+This is an active master with a standard UVM structure (sequencer, driver, monitor).
+Code implementing the agent can be found in [`dv/uvm/icache/dv/ibex_icache_core_agent`](https://github.com/lowRISC/ibex/tree/master/dv/uvm/icache/dv/ibex_icache_core_agent).
+
+The complicated internal state of the cache is somewhat visible to the core, resulting in strict rules about how the core drives it (see the [detailed behaviour](https://ibex-core.readthedocs.io/en/latest/icache.html#detailed-behaviour) section of the documentation).
+These rules actually make the cache easier to verify because they constrain how seemingly unrelated signals (`req_i`, `icache_inval_i`, `icache_enable_i`) must be driven relative to instruction requests, meaning a simpler set of valid transactions.
+These transactions are:
+
+ * Branch to address `A`, maybe invalidating and/or enabling/disabling the cache, and read `N ≥ 0` instructions. Stop after fewer instructions if there's an error.
+ * Disable `req_i` for `C ≥ 0` cycles, possibly invalidating and/or enabling/disabling the cache.
+   When it comes back up, read `N ≥ 0` instructions, stopping early on error.
+
+The monitor for the core agent reports cache invalidations, when the cache is enabled/disabled, branches and instruction reads (including errors).
+
+#### Memory Agent
+
+The memory agent emulates the instruction bus, supplying data to the cache.
+This must be deterministic (in order for the scoreboard to tell whether the cache fetched the right data), but must also be able to change with time (in order to check invalidation works correctly).
+To support this, the memory contents are modelled with a 32-bit seed value.
+
+The architectural state of the instruction bus and memory (contents at each address; PMP ranges; ranges that will cause a memory error) are all derived from this seed.
+The precise functions can be found in [`dv/uvm/icache/dv/ibex_icache_mem_agent/ibex_icache_mem_model.sv`](https://github.com/lowRISC/ibex/blob/master/dv/uvm/icache/dv/ibex_icache_mem_agent/ibex_icache_mem_model.sv).
+
+The memory agent is an active slave, responding to instruction fetches from the cache with either a PMP error (on the same cycle as the request) or instruction data (with an in-order request pipeline).
+As well as the instruction fetch requests, its sequence contains occasional transactions that change the underlying seed.
```

Comment:
It sounds like when the seed is first set, the memory model (plus pmp regions, etc...) will be randomly generated based on this value, correct? Does this then mean that when the cache is invalidated due to `fence.i` reset the entire memory model will be re-generated due to the seed of the memory model changing?

Reply:
An invalidation doesn't change the memory seed (after all: in the real system, an invalidation is different from a memory write).

The idea is that the memory seed changes occasionally, asynchronously from what the core is doing. The memory sequence will inject occasional "change the seed" transactions and the scoreboard will check that results from fetches match at least one of the seeds that has been seen since the last invalidation. Incidentally, the "entire memory model" is just a 32-bit integer: everything else is calculated with hash functions from that seed.

Is there a bit of this text that is confusing? It makes sense to me, but then again I wrote it, so I'm probably not a very good judge!

Reply:
Ah I understand now, I just had some confusion whether the seed would be changed asynchronously from the core sending an invalidation request or not.

To clarify something in your comment, if data in the "memory model" is calculated based off of a hash of the seed, will it be possible for the result of an instruction fetch to match more than 1 seed?

Reply:
No: it will match exactly one seed. But if the cache is enabled (and we're not modelling eviction strategies etc.), there are a set of seeds that it might have come from (I'll get the old version of the data if cached; the new version if not).

Reply:
Sure, that makes sense to me, thanks for the clarification.